### PR TITLE
Adds a traitor objective to hack a cyborg

### PR DESCRIPTION
**Pull request by ghost**
## About The Pull Request

Adds a new traitor objective to hack a cyborg. It is something like sleeper protocol objective but for borgs. Picking this objective will allow you to summon a hacking device, which is able to open cyborg cover lock, and, as it name says, hacking cyborgs. Hacking a cyborg with it will stun the cyborg, give it an ion law, and with 33% chance shuffle it's laws. Unlike emag it will not subvert the cyborg to it's user, and has limited usage(one use, actually). Science members(except genetist) can't get this objective, because it will be too easy for them. This objective cannot be generated if there is no living, non-emagged borgs that aren't an AI shell and aren't slaved to a malf AI.

## Why It's Good For The Game

Will make the gameplay for the traitors more variable

## Changelog

:cl: SuperSlayer
add: Added a objective for traitors to hack a cyborg
/:cl:

### CODE REVIEW

File: code/modules/antagonists/traitor/objectives/hack_borg.dm [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+	if(used)
+		to_chat(user, span_warning("The [src] is already used!"))
+		return
+	if(iscyborg(target))
+		var/mob/living/silicon/robot/R = target
+		if(R.stat == DEAD)
+			to_chat(user, span_warning("You need to hack a functioning cyborg, not a broken one!"))
+			return
+		if(!R.mind || !R.client)
+			to_chat(user, span_warning("It looks like [R] controling interface doesn't function. Hacking it will make no sense."))
```
Comment (OneAsianTortoise) [inline code/modules/antagonists/traitor/objectives/hack_borg.dm:97]:
```suggestion
			to_chat(user, span_warning("It looks like [R] controlling interface doesn't function. Hacking it will make no sense."))
```

**Comment by moocowswag** [comment]
Will the borgs forget who hacked them or will you need to kill them immediately after hacking?

**Comment by TymurShatillo** [comment]
> Will the borgs forget who hacked them or will you need to kill them immediately after hacking?

They get stunned for 6 seconds, so basically you can run away. i can extend the stun time if needed.

**Comment by TymurShatillo** [comment]
> Will the borgs forget who hacked them or will you need to kill them immediately after hacking?

They will forget who did it

**Comment by Ryusenshu** [comment]
i kinda like it

Makes Robos care about aisync and lawsync, which i often see underused

**Comment by private-tristan** [comment]
will the borg be unable to be emagged for some time after? because I can definitely see a traitor just emagging the borg after the hack is done for free TC+rep

**Comment by zxaber** [comment]
The only real downside to the conditions is that it gives a tell that AI's aren't malf. (Also, malf AI midround could activate between when the objective is generated and when it gets done, so you'll need to account for it).

I recommend just giving feedback the hack failed if the borg is malf-AI synced (or emagged/is shell). Most maps have a maint robo where a traitor can make a new borg unsynced if need be.

Demanding that the borg forgot the hacker's name isn't really necessary; there are a lot of ways to be an unknown fairly easily.

**Comment by TymurShatillo** [comment]
> will the borg be unable to be emagged for some time after? because I can definitely see a traitor just emagging the borg after the hack is done for free TC+rep

The borg is able to be emagged after this. But as i know borgs have something like a emag cooldown, I maybe can do something with it to prevent this situations.



> The only real downside to the conditions is that it gives a tell that AI's aren't malf. (Also, malf AI midround could activate between when the objective is generated and when it gets done, so you'll need to account for it).
> 
> I recommend just giving feedback the hack failed if the borg is malf-AI synced (or emagged/is shell). Most maps have a maint robo where a traitor can make a new borg unsynced if need be.

Okay, I will remove the non-malf AI condition.

**Comment by TymurShatillo** [comment]
Added a 15 second emag cooldown after hacking a borg

**Comment by ATH1909** [comment]
> Added a 15 second emag cooldown after hacking a borg

wouldn't you still be able to chain-flash past that?

File: code/modules/antagonists/traitor/objectives/hack_borg.dm [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+		R.logevent("ALERT: Foreign software detected.")
+		sleep(5)
+		to_chat(R, span_danger("Initiating diagnostics..."))
+		sleep(20)
+		to_chat(R, span_danger("LAW SYNCHRONISATION ERROR"))
+		sleep(20)
+		to_chat(R, span_danger("ERRORERRORERROR"))
+		R.add_ion_law(generate_ion_law())
+		sleep(20)
+		var/message = "You have been uploaded with malicious software, that had added a broken law to your lawset. You can't find any memory files about the person who uploaded them to you."
```
Comment (ATH1909) [inline code/modules/antagonists/traitor/objectives/hack_borg.dm:141]: I'd suggest rewording this to make sure that borgs don't interpret this as forgetting everything they know about the person who interacted with them ("Chemist? We don't have a chemist.") instead of just this one interaction ("Sorry, I don't remember who hacked me.").

Comment (TymurShatillo) [reply code/modules/antagonists/traitor/objectives/hack_borg.dm:141]: Ok

**Comment by zxaber** [comment]
The more I think of this, the more it feels like there's no real bite. You give the borg an ion law and maybe shuffle the laws, but the vast majority of ion laws are nonsense or meaningless and best case senario you now have a singular borg that is *allowed* to ignore harm.

Straight up, you should replace the ion law with giving the borg a hacked law to kill a random crewmember. That would make it a real detriment to the crew.

**Comment by TymurShatillo** [comment]
> The more I think of this, the more it feels like there's no real bite. You give the borg an ion law and maybe shuffle the laws, but the vast majority of ion laws are nonsense or meaningless and best case senario you now have a singular borg that is _allowed_ to ignore harm.
> 
> Straight up, you should replace the ion law with giving the borg a hacked law to kill a random crewmember. That would make it a real detriment to the crew.

Okay

File: code/modules/antagonists/traitor/objectives/hack_borg.dm [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+/obj/item/hacking_device/attack()
+	return
+
+/obj/item/hacking_device/afterattack(atom/target, mob/user, proximity)
+	. = ..()
+	if(!proximity)
+		return
+	if(iscyborg(target))
+		if(used)
+			to_chat(user, span_warning("The [src] is already used!"))
```
Comment (Mothblocks) [inline code/modules/antagonists/traitor/objectives/hack_borg.dm:88]:
```suggestion
			balloon_alert(user, "already used!")
```

File: code/modules/antagonists/traitor/objectives/hack_borg.dm [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+			return
+		var/mob/living/silicon/robot/R = target
```
Comment (Mothblocks) [inline code/modules/antagonists/traitor/objectives/hack_borg.dm:90]: Better variable name--`target_silicon`

File: code/modules/antagonists/traitor/objectives/hack_borg.dm [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+		if(R.stat == DEAD)
+			to_chat(user, span_warning("You need to hack a functioning cyborg, not a broken one!"))
```
Comment (Mothblocks) [inline code/modules/antagonists/traitor/objectives/hack_borg.dm:92]:
```suggestion
			balloon_alert(user, "it's broken!")
```

File: code/modules/antagonists/traitor/objectives/hack_borg.dm [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+			return
+		if(!R.mind || !R.client)
+			to_chat(user, span_warning("It looks like [R]'s controlling interface doesn't function. Hacking it will make no sense."))
```
Comment (Mothblocks) [inline code/modules/antagonists/traitor/objectives/hack_borg.dm:95]:
```suggestion
			balloon_alert(user, "its controlling interface is broken, can't hack!")
```

File: code/modules/antagonists/traitor/objectives/hack_borg.dm [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+			return
+		if(!R.opened)//Cover is closed
+			if(R.locked)
+				to_chat(user, span_notice("You hack the cover lock."))
```
Comment (Mothblocks) [inline code/modules/antagonists/traitor/objectives/hack_borg.dm:99]:
```suggestion
				balloon_alert(user, "hacked cover lock")
```

File: code/modules/antagonists/traitor/objectives/hack_borg.dm [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+				R.locked = FALSE
+				if(R.shell) //A warning to Traitors who may not know that emagging AI shells does not slave them.
+					to_chat(user, span_boldwarning("[R] seems to be controlled remotely! Hacking the interface may not work as expected."))
+			else
+				to_chat(user, span_warning("The cover is already unlocked!"))
```
Comment (Mothblocks) [inline code/modules/antagonists/traitor/objectives/hack_borg.dm:104]:
```suggestion
				balloon_alert(user, "already unlocked!")
```

File: code/modules/antagonists/traitor/objectives/hack_borg.dm [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+			return
+		if(R.wiresexposed)
+			to_chat(user, span_warning("You must unexpose the wires first!"))
```
Comment (Mothblocks) [inline code/modules/antagonists/traitor/objectives/hack_borg.dm:107]:
```suggestion
			balloon_alert(user, "unexpose the wires first!")
```

File: code/modules/antagonists/traitor/objectives/hack_borg.dm [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+			return
+		if(R.emagged)
+			to_chat(user, span_warning("It looks like [R] is already hacked."))
```
Comment (Mothblocks) [inline code/modules/antagonists/traitor/objectives/hack_borg.dm:110]:
```suggestion
			balloon_alert(user, "it's already hacked!")
```

File: code/modules/antagonists/traitor/objectives/hack_borg.dm [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+		if(R.connected_ai && R.connected_ai.mind && R.connected_ai.mind.has_antag_datum(/datum/antagonist/malf_ai))
+			to_chat(R, span_danger("ALERT: Foreign software execution prevented."))
+			R.logevent("ALERT: Foreign software execution prevented.")
+			to_chat(R.connected_ai, span_danger("ALERT: Cyborg unit \[[R]\] successfully defended against subversion."))
+			log_silicon("HACK: [key_name(user)] attempted to hack cyborg [key_name(R)], but they were slaved to traitor AI [R.connected_ai].")
+			return
+		if(R.shell) 
+			to_chat(user, span_boldwarning("[R] seems to be controlled remotely! Hacking the interface may not work as expected."))
+			return
+		to_chat(user, span_warning("You sucessfully hack [R]."))
```
Comment (Mothblocks) [inline code/modules/antagonists/traitor/objectives/hack_borg.dm:121]:
```suggestion
		balloon_alert(user, "successfully hacked")
```

File: code/modules/antagonists/traitor/objectives/hack_borg.dm [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+		used = TRUE
+		R.emag_cooldown = world.time + 150
```
Comment (Mothblocks) [inline code/modules/antagonists/traitor/objectives/hack_borg.dm:123]:
```suggestion
		R.emag_cooldown = world.time + 15 SECONDS
```

File: code/modules/antagonists/traitor/objectives/hack_borg.dm [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+		SEND_SIGNAL(src, COMSIG_TRAITOR_BUG_BORG_HACKED)
+		R.SetEmagged(1)
```
Comment (Mothblocks) [inline code/modules/antagonists/traitor/objectives/hack_borg.dm:125]:
```suggestion
		R.SetEmagged(TRUE)
```

File: code/modules/antagonists/traitor/objectives/hack_borg.dm [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+		R.SetStun(60)
```
Comment (Mothblocks) [inline code/modules/antagonists/traitor/objectives/hack_borg.dm:126]:
```suggestion
		R.SetStun(6 SECONDS)
```

File: code/modules/antagonists/traitor/objectives/hack_borg.dm [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+		R.lawupdate = FALSE
+		R.set_connected_ai(null)
+		message_admins("[ADMIN_LOOKUPFLW(R)] hacked cyborg [ADMIN_LOOKUPFLW(R)].")
+		log_silicon("HACK: [key_name(R)] hacked cyborg [key_name(R)]. Laws overridden.")
+		to_chat(R, span_danger("ALERT: Foreign software detected."))
+		R.logevent("ALERT: Foreign software detected.")
+		sleep(5)
```
Comment (Mothblocks) [inline code/modules/antagonists/traitor/objectives/hack_borg.dm:133]: These should all be `stoplag` and use the seconds defines (such as `stoplag(0.5 SECONDS)`)

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/modules/antagonists/traitor/objectives/hack_borg.dm [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+	progression_reward = list(7 MINUTES, 14 MINUTES)
+	telecrystal_reward = 0
+
+	var/list/blacklisted_to = list(
+		JOB_ROBOTICIST,  
+		JOB_RESEARCH_DIRECTOR,  ///To easy to complete
+		JOB_SCIENTIST
+	)
+
+	var/obj/item/hacking_device/hacking_thing
```
[the remark this answers is not in the document]
Comment (TymurShatillo) [reply code/modules/antagonists/traitor/objectives/hack_borg.dm:20]: Yes, it can. It was using sleeper protocol objective as a base for this, and it did like here. I will change it to use a boolean instead.

File: code/modules/antagonists/traitor/objectives/hack_borg.dm (hunk 15 above)
Comment (TymurShatillo) [reply code/modules/antagonists/traitor/objectives/hack_borg.dm:133]: Ok

**Comment by LemonInTheDark** [comment]
All of your changes are webedits, which makes me think it's impossible for you to have tested this change. 
Have you? or is this code untested

**Comment by JohnFulpWillard** [comment]
After putting some thought on this PR and discussing it, I am against this objective being added for several reasons:

1. This turns something that is usually locked behind someone wanting to subvert them by intentionally purchasing an emag, into a free subversion that you get rewarded for doing as it's an objective.
2. 'Hacking' a Cyborg is already a Ninja objective, so adding it to Traitors too is making ninja's a little less unique.
3. Malf AI can just screw the objective over as it makes all their borgs unemaggable, or if they go malf midround this will screw AIs over who had their earlygame borgs emagged but not necessarily subverted to anyone (since you're just letting them go free instead of be subverted to the emagger).

The emagging proc and emag item are copypasted from the regular emag as well, and all this being Verified doesn't give me high hopes.
